### PR TITLE
pyknowhere - support dump and load index file

### PR DESCRIPTION
**Pull request by alwayslove2013**
/kind feature

Issue: #154

### CODE REVIEW

File: python/knowhere/knowhere.i [hunk 1]
```diff
@@ -354,4 +356,51 @@ DumpRangeResultDis(knowhere::DataSetPtr result, float* dis, int len) {
     }
 }
 
+void
+Dump(knowhere::BinarySetPtr binset, const std::string& file_name) {
+    auto binary_set = *binset;
+    auto binary_map = binset -> binary_map_;
+    std::ofstream outfile;
+    outfile.open(file_name, std::ios::out | std::ios::trunc);
+    for (auto it = binary_map.begin(); it != binary_map.end(); ++it) {
```
Comment (alexanderguzhva) [inline python/knowhere/knowhere.i:365]: please check is the file was opened successfully via `if (outfile.good()) {}`

File: python/knowhere/knowhere.i [hunk 2]
```diff
@@ -354,4 +356,51 @@ DumpRangeResultDis(knowhere::DataSetPtr result, float* dis, int len) {
...
+    // end with 0
+    outfile << 0;
+    outfile.close();
+}
+
+void
+Load(knowhere::BinarySetPtr binset, const std::string& file_name) {
+    std::ifstream infile;
+    infile.open(file_name, std::ios::in);
+    size_t name_len;
```
Comment (alexanderguzhva) [inline python/knowhere/knowhere.i:383]: please check is the file was opened via `if (infile.good()) {}`

File: python/knowhere/knowhere.i [hunk 3]
```diff
@@ -354,4 +356,51 @@ DumpRangeResultDis(knowhere::DataSetPtr result, float* dis, int len) {
...
+Dump(knowhere::BinarySetPtr binset, const std::string& file_name) {
+    auto binary_set = *binset;
+    auto binary_map = binset -> binary_map_;
+    std::ofstream outfile;
+    outfile.open(file_name, std::ios::out | std::ios::trunc);
+    for (auto it = binary_map.begin(); it != binary_map.end(); ++it) {
+        // serialization: name_length(size_t); name(char[]); binset_size(size_t); binset(uint8[]); 
+        auto name = it->first;
+        outfile << name.size();
+        outfile << name;
```
Comment (alexanderguzhva) [inline python/knowhere/knowhere.i:369]: name.size() is type `size_t` and is platform dependent, please write sizes as `int64_t`

File: python/knowhere/knowhere.i [hunk 4]
```diff
@@ -354,4 +356,51 @@ DumpRangeResultDis(knowhere::DataSetPtr result, float* dis, int len) {
...
+        if (name_len == 0) break;
+
+        auto _name = new char[name_len];
+        infile.read(_name, name_len);
+        std::string name(_name, name_len);
+
+        int64_t size;
+        infile >> size;
+
+        auto data = new uint8_t[size];
```
Comment (alexanderguzhva) [inline python/knowhere/knowhere.i:397]: any checks if the size is too large, negative or making no sense? :)

File: python/knowhere/knowhere.i [hunk 5]
```diff
@@ -354,4 +356,51 @@ DumpRangeResultDis(knowhere::DataSetPtr result, float* dis, int len) {
...
+        infile.read(reinterpret_cast<char*>(data), size);
+        std::shared_ptr<uint8_t[]> data_ptr(data);
```
Comment (alexanderguzhva) [inline python/knowhere/knowhere.i:399]: `std::shared_ptr<>` first, read data into it next

File: python/knowhere/knowhere.i [hunk 6]
```diff
@@ -354,4 +356,51 @@ DumpRangeResultDis(knowhere::DataSetPtr result, float* dis, int len) {
...
+        auto name = it->first;
+        outfile << name.size();
+        outfile << name;
+        auto value = it->second;
+        outfile << value->size;
+        outfile.write(reinterpret_cast<char*>(value->data.get()), value->size);
+    }
+    // end with 0
+    outfile << 0;
+    outfile.close();
```
Comment (alexanderguzhva) [inline python/knowhere/knowhere.i:376]: not needed.

File: python/knowhere/knowhere.i [hunk 7]
```diff
@@ -354,4 +356,51 @@ DumpRangeResultDis(knowhere::DataSetPtr result, float* dis, int len) {
...
+        int64_t size;
+        infile >> size;
+
+        auto data = new uint8_t[size];
+        infile.read(reinterpret_cast<char*>(data), size);
+        std::shared_ptr<uint8_t[]> data_ptr(data);
+
+        binset->Append(name, data_ptr, size);
+    }
+    infile.close();
```
Comment (alexanderguzhva) [inline python/knowhere/knowhere.i:403]: `infile.close();` is not needed

File: python/knowhere/knowhere.i (hunk 1 above)
Comment (alwayslove2013) [reply python/knowhere/knowhere.i]: fixed

File: python/knowhere/knowhere.i (hunk 2 above)
Comment (alwayslove2013) [reply python/knowhere/knowhere.i]: fixed

File: python/knowhere/knowhere.i (hunk 3 above)
Comment (alwayslove2013) [reply python/knowhere/knowhere.i]: fixed

File: python/knowhere/knowhere.i (hunk 5 above)
Comment (alwayslove2013) [reply python/knowhere/knowhere.i]: how about this
`std::shared_ptr<uint8_t[]> data_ptr(data);`
`infile.read(reinterpret_cast<char*>(data_ptr.get()), size);`

File: python/knowhere/knowhere.i (hunk 4 above)
Comment (alwayslove2013) [reply python/knowhere/knowhere.i]: negative checked.
how to check for too large size? or we can just wait for the program to report an exception?

File: python/knowhere/knowhere.i (hunk 6 above)
Comment (alwayslove2013) [reply python/knowhere/knowhere.i]: fixed

File: python/knowhere/knowhere.i (hunk 7 above)
Comment (alwayslove2013) [reply python/knowhere/knowhere.i]: fixed

File: python/knowhere/knowhere.i (hunk 5 above)
Comment (alexanderguzhva) [reply python/knowhere/knowhere.i]: yes. 
Also, please note that this code needs C++17, check https://stackoverflow.com/questions/13061979/can-you-make-a-stdshared-ptr-manage-an-array-allocated-with-new-t , I don't know how this shared_ptr<uint8_t[]> will be used later.

**Comment by alexanderguzhva** [comment]
It needs to be noted somewhere that reading/writing errors are not exposed

**Comment by alwayslove2013** [comment]
@alexanderguzhva The load and dump API changes will only appear in `pyknowhere`, and will not participate in the compilation of knowhere c++. How to accurately expose error messages to Python is still rough in the current version, and seems to need to be redesigned.

By the way, I have squashed all commits into one, and `LGTM` label has been removed. Please give another one~

**Comment by liliu-z** [comment]
@alwayslove2013 issue link plz

**Comment by alwayslove2013** [comment]
related issue: #154

**Pull request merged**